### PR TITLE
Incprporate changes in ZigZag after New User Interface PR

### PR DESCRIPTION
**Pull request by RobinGeens**
Make Stream compatible with the new ZigZag version from the pull request [New User Interface](https://github.com/KULeuven-MICAS/zigzag/pull/50). The code of this PR will only work once the ZigZag version has been upgraded and the changes from the New User Interface PR are included in the package.

[New User Interface](https://github.com/KULeuven-MICAS/zigzag/pull/50) no longer works for IMC architectures and ZigZag will need to be upgraded further to once again support IMC. Doing so will most likely also require an update in Stream.

### CODE REVIEW

File: stream/classes/hardware/architecture/noc/communication_link.py [hunk 1]
```diff
@@ -6,9 +6,7 @@
 class CommunicationLink:
     """Represents a fixed-bandwidth communication link used to communicate between two cores."""
 
-    def __init__(
-        self, sender, receiver, bandwidth, unit_energy_cost, bidirectional=False
-    ) -> None:
+    def __init__(self, sender, receiver, bandwidth, unit_energy_cost, bidirectional=False) -> None:
```
Comment (Andful) [inline stream/classes/hardware/architecture/noc/communication_link.py:9]: Saw that in many places that the argument type is being introduced. Was this supposed to happen here too?

probably could be something along the lines of
`def __init__(self, sender: int | Literal["Any"], receiver: int | Literal["Any"], bandwidth: int, unit_energy_cost: int, bidirectional: bool=False)`

Comment (RobinGeens) [reply stream/classes/hardware/architecture/noc/communication_link.py:9]: Adding type hints is a WIP and I wasn't able to do it everywhere in Stream. They can definitely be added to other classes and functions in the future.

Not sure what the point is of adding `Literal["Any"]` as a type hint? Won't this accept any argument?

The main benefit of type hinting the whole code base is that you can run pytest with the typeguard plugin, which dynamically type checks all function calls.

File: stream/visualization/schedule.py [hunk 2]
```diff
@@ -38,17 +39,17 @@
...
+    section_start_percent: tuple[int, ...] = (0, 50, 95),
+    percent_shown: tuple[int, ...] = (5, 5, 5),
+    plot_data_transfer: bool = False,
+    fig_path: str = "outputs/schedule_plot.png",
+) -> None:
+    G: Workload = scme.workload
     accelerator: Accelerator = scme.accelerator
 
-    nb_layers = len(set(iter([n.id[0] for n in G.nodes()])))
+    nb_layers = len(set(iter([n.id for n in G.nodes()])))
```
Comment (Andful) [inline stream/visualization/schedule.py:52]: Does this not change the logical meaning of the code?

Stream uses a tuple id of the form `(layer, chunk)`. In the case of layer fusion, `layer` is ... the layer... and `chunk` is the sub-layer split.

The code previously only checked for the number of layers but now it effectively checks for the number of computational nodes

Comment (Andful) [reply stream/visualization/schedule.py:52]: Ah nvm, I notice now that there is a `sub_id` field

File: stream/classes/stages/InterCoreMappingStage.py [hunk 3]
```diff
@@ -25,18 +32,18 @@ class InterCoreMappingStage(Stage):
...
-        accelerator,
-        node_hw_performances,
-        nb_ga_generations,
-        nb_ga_individuals,
-        plot_hof,
-        plot_file_name,
-        plot_full_schedule=False,
-        plot_data_transfer=False,
-        operands_to_prefetch,
+        workload: Workload,
```
Comment (Andful) [inline stream/classes/stages/InterCoreMappingStage.py:37]: Should this be a `Workload`? Is it not a `DiGraph`? from my understanding, after the workload is split into computational nodes, it stops being a `Workload` and it becomes `DiGraph`.

File: .gitmodules [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[submodule "zigzag_repo/zigzag"]
```
Comment (Andful) [inline .gitmodules:1]: Should this be here?

Comment (asyms) [reply .gitmodules:1]: I don't think so. I think @RobinGeens used this for testing?

File: stream/classes/cost_model/cost_model.py [hunk 5]
```diff
@@ -14,10 +15,10 @@ class StreamCostModelEvaluation:
 
     def __init__(
         self,
-        workload: DiGraph,
+        workload: Workload,
         accelerator: Accelerator,
-        operands_to_prefetch: list,
-        scheduling_order: list,
+        operands_to_prefetch: list[str],
```
Comment (Andful) [inline stream/classes/cost_model/cost_model.py:20]: Not sure, but should this not be `MemoryOperand`?

File: stream/classes/cost_model/scheduler.py [hunk 6]
```diff
@@ -234,7 +211,7 @@ def check_for_removal(
 def schedule_graph(
     G: DiGraph,
```
Comment (Andful) [inline stream/classes/cost_model/scheduler.py:212]: Is it fine that here is `DiGraph` and not `Workload`?

File: stream/classes/stages/InterCoreMappingStage.py (hunk 3 above)
Comment (Andful) [reply stream/classes/stages/InterCoreMappingStage.py:37]: Aaah, thank you for the clarification!

File: stream/classes/hardware/architecture/noc/communication_link.py (hunk 1 above)
Comment (RobinGeens) [reply stream/classes/hardware/architecture/noc/communication_link.py:9]: Ahh I see. I think `Core` and `int` are both used throughout the code.

File: .gitmodules (hunk 4 above)
Comment (RobinGeens) [reply .gitmodules]: This should not be here. Fixed :)

File: stream/classes/cost_model/cost_model.py (hunk 5 above)
Comment (RobinGeens) [reply stream/classes/cost_model/cost_model.py:20]: Good catch, but no. Almost all user input is now provided through the yaml files, and memory operand names are parsed as `MemoryOperand`. The `operands_to_prefetch` is provided in the `main.py` file itself. I didn't want to let the user convert it himself in the definition of  `operands_to_prefetch`, so it is converted to the correct class within the function that uses ``operands_to_prefetch`.

File: stream/classes/cost_model/scheduler.py (hunk 6 above)
Comment (RobinGeens) [reply stream/classes/cost_model/scheduler.py:212]: Both work, I guess I forgot to change this one to `Workload`. Note that the type hints are all but finished, and don't really matter that much for the moment (they are not actually checked).

**Review by asyms — APPROVED** [review]
LGTM

**Pull request merged**